### DISCUSSION TITLE
The expressionsem file have a code smell.

### DISCUSSION THREAD

**Discussion by 12345swordy** [dlang/dmd#13080] [General]
It is literally over 10k lines of code, the majority of it consist of function overloading for different classes. Why can't there be files for every class that has functions being overloaded?

**Comment by UplinkCoder** [comment]
Because that doesn't make anything cleaner it just spreads it around the
filesystem.

Am Fr., 17. Sept. 2021 um 21:18 Uhr schrieb 12345swordy <
***@***.***>:

**Reply by 12345swordy** [reply]
Surely, you have better ideas then the current status quo? There is now semantic analysis for import C in the expressionsem file now, which is not obvious to anyone who is new to the compiler internals.

**Comment by maxhaton** [comment]
Breaking it up into files requires sensible places for the files to go (packages) and [that did not end well last time or the time before that or the time before that or the time before that etc. etc.. ](https://youtu.be/MniVeP3HIDA) 

The current file-osophy is akin to filling a big tray of shot glasses with a garden hose - you fill up the ones you aim for but the fluid still spills all over the place e.g.`templateInstanceSemantic` is buried in the middle of nowhere in dsymbolsem IIRC. It could be worse in that this is somewhat preferable to "Enterprise" code which actively hides the interesting stuff, but even GCC is downright poetry compared to us in some ways and GCC has way more baggage and a vastly inferior emergent-language (due to C -> C++11 transition it's still somewhat in between the two)
